### PR TITLE
fix(model-datastructure): make `MapBasedStore` thread safe

### PR DESCRIPTION
**Pull request by odzhychko**
Make it thread safe by using a synchronized map.

Missing thread safety is the suspected issue for such error messages

```
2024-03-20 08:58:01,282 [  48184]  ERROR - #com.intellij.ide.plugins.PluginManager - Equal objects must have equal hashcodes. During rehashing, Trove discovered that the following two objects claim to be equal (as in java.lang.Object.equals()) but their hashCodes (or those calculated by your TObjectHashingStrategy) are not equal.This violates the general contract of java.lang.Object.hashCode().  See bullet point two in that method's documentation. object #1 =class java.lang.String id= 996201210 hashCode= -209607169 toString= sHB6r*02gLuXVEvI31E7v9ijDVL5VSdSO8F9OO-2z2CA; object #2 =class java.lang.String id= 1530156665 hashCode= -209607169 toString= sHB6r*02gLuXVEvI31E7v9ijDVL5VSdSO8F9OO-2z2CA
[Warning] apparent concurrent modification of the key set. Size before and after rehash() do not match 81 vs 83
java.lang.IllegalArgumentException: Equal objects must have equal hashcodes. During rehashing, Trove discovered that the following two objects claim to be equal (as in java.lang.Object.equals()) but their hashCodes (or those calculated by your TObjectHashingStrategy) are not equal.This violates the general contract of java.lang.Object.hashCode().  See bullet point two in that method's documentation. object #1 =class java.lang.String id= 996201210 hashCode= -209607169 toString= sHB6r*02gLuXVEvI31E7v9ijDVL5VSdSO8F9OO-2z2CA; object #2 =class java.lang.String id= 1530156665 hashCode= -209607169 toString= sHB6r*02gLuXVEvI31E7v9ijDVL5VSdSO8F9OO-2z2CA
[Warning] apparent concurrent modification of the key set. Size before and after rehash() do not match 81 vs 83
	at gnu.trove.impl.hash.TObjectHash.buildObjectContractViolation(TObjectHash.java:464)
	at gnu.trove.impl.hash.TObjectHash.throwObjectContractViolation(TObjectHash.java:448)
	at gnu.trove.map.hash.THashMap.rehash(THashMap.java:403)
	at gnu.trove.impl.hash.THash.ensureCapacity(THash.java:175)
	at gnu.trove.map.hash.THashMap.putAll(THashMap.java:551)
	at org.modelix.model.persistent.MapBasedStore.putAll(MapBasedStore.kt:54)
```

### CODE REVIEW

**Comment by odzhychko** [comment]
> Could this again become an issue with threading primitives being used in coroutine contexts later on?

I am not sure if I understand you correctly. Maybe my answer still helps.

The synchronized map ensures that two threads do not use the map at the same time.
When a coroutine is executed, it is executed on a thread.
Therefore, two coroutines running at the same time on different threads will not use the synchronized map at the same time.

**Comment by languitar** [comment]
I was wondering if introducing more threading locks can lead to a similar situation fixed with https://github.com/modelix/modelix.core/pull/558. Maybe usage patterns should once be checked.

**Comment by odzhychko** [comment]
`MapBasedStore` is only used in the `ModelClientV2`. So this change will not affect the server.

**Comment by languitar** [comment]
image

Maybe we should configure cucumber to not include the chrome version in the test name?

**Comment by odzhychko** [comment]
> Maybe we should configure cucumber to not include the chrome version in the test name?

I think this is caused by Kotlin-Multiplatform and not Cucumber.
I created a ticket for that issue https://issues.modelix.org/issue/MODELIX-836

File: kotlin-utils/src/commonMain/kotlin/org/modelix/kotlin/utils/DataStructures.kt [hunk 1]
```diff
@@ -29,3 +29,5 @@ package org.modelix.kotlin.utils
  * Therefore, the memory efficient maps are used sparingly for only the very big maps.
  */
 expect fun <K, V> createMemoryEfficientMap(): MutableMap<K, V>
+
+expect fun <K, V> synchronizedMap(map: MutableMap<K, V>): MutableMap<K, V>
```
Comment (slisson) [inline kotlin-utils/src/commonMain/kotlin/org/modelix/kotlin/utils/DataStructures.kt:33]: `fun <K, V> MutableMap<K, V>.toSynchronizedMap(): MutableMap<K, V>` might be more idiomatic Kotlin code.

**Review by slisson — APPROVED** [review]

**Pull request merged**